sim/druid/druid.go

```diff
@@ -3,30 +3,201 @@ package druid
+	SelfBuffs
+	Talents proto.DruidTalents
+
+	innervateCD  time.Duration
+	naturesGrace bool // when true next spellcast is 0.5s faster
+
+	// cached cast stuff
+	starfireSpell         core.SingleTargetDirectDamageSpell
+	starfire8CastTemplate core.SingleTargetDirectDamageSpellTemplate
+	starfire6CastTemplate core.SingleTargetDirectDamageSpellTemplate
+
+	MoonfireSpell        core.DamageOverTimeSpell
+	moonfireCastTemplate core.DamageOverTimeSpellTemplate
+
+	wrathSpell        core.SingleTargetDirectDamageSpell
+	wrathCastTemplate core.SingleTargetDirectDamageSpellTemplate
+
+	InsectSwarmSpell        core.DamageOverTimeSpell
+	insectSwarmCastTemplate core.DamageOverTimeSpellTemplate
+}
+
+type SelfBuffs struct {
+	Omen      bool
+	Innervate bool
 }
 
 func (druid *Druid) GetCharacter() *core.Character {
 	return &druid.Character
 }
 
 func (druid *Druid) AddRaidBuffs(raidBuffs *proto.RaidBuffs) {
-	// TODO: Use talents to check for imp gotw
 	raidBuffs.GiftOfTheWild = proto.TristateEffect_TristateEffectRegular
+	if druid.Talents.ImprovedMarkOfTheWild > 0 { // ya ya whatever
+		raidBuffs.GiftOfTheWild = proto.TristateEffect_TristateEffectImproved
+	}
 }
+
+const ravenGoddessItemID = 32387
+
 func (druid *Druid) AddPartyBuffs(partyBuffs *proto.PartyBuffs) {
-	//buffs.Moonkin = proto.TristateEffect_TristateEffectRegular
-	// check for idol of raven goddess equipped
+	if druid.Talents.MoonkinForm { // assume if you have moonkin talent you are using it.
+		partyBuffs.MoonkinAura = proto.TristateEffect_TristateEffectRegular
+		for _, e := range druid.Equip {
+			if e.ID == ravenGoddessItemID {
+				partyBuffs.MoonkinAura = proto.TristateEffect_TristateEffectImproved
+				break
+			}
+		}
+	}
+}
+
+func (druid *Druid) Init(sim *core.Simulation) {
+	druid.starfire8CastTemplate = druid.newStarfireTemplate(sim, 8)
+	druid.starfire6CastTemplate = druid.newStarfireTemplate(sim, 6)
+	druid.moonfireCastTemplate = druid.newMoonfireTemplate(sim)
+	druid.wrathCastTemplate = druid.newWrathTemplate(sim)
+	druid.insectSwarmCastTemplate = druid.newInsectSwarmTemplate(sim)
 }
 
 func (druid *Druid) Reset(newsim *core.Simulation) {
+	druid.moonfireCastTemplate.Apply(&druid.MoonfireSpell)
+	druid.Character.Reset(newsim)
 }
 
+func (druid *Druid) Advance(sim *core.Simulation, elapsedTime time.Duration) {
+	druid.Character.RegenManaMP5Only(sim, elapsedTime)
+	druid.Character.Advance(sim, elapsedTime)
+}
+
+var InnervateCD = core.NewCooldownID()
+var InnervateAuraID = core.NewAuraID()
+
+func (druid *Druid) TryInnervate(sim *core.Simulation) {
+	// Currently just activates innervate on self when own mana is <75%
+	// TODO: get a real recommendation when to use this.
+	if druid.SelfBuffs.Innervate && druid.GetRemainingCD(InnervateCD, sim.CurrentTime) == 0 {
+		if druid.GetStat(stats.Mana)/druid.MaxMana() < 0.75 {
+			oldRegen := druid.PsuedoStats.SpiritRegenRateCasting
+			druid.PsuedoStats.SpiritRegenRateCasting = 1.0
+			druid.PsuedoStats.ManaRegenMultiplier *= 3.0
```

Comment:
Innervate is INCREASE by 400% so I think this is x5? Not sure exactly how this is assembled though.

Reply:
oh ya, math